### PR TITLE
feat(core): add `distPath.assets`  and `filename.assets`

### LINKED ISSUE #3650: [Bug]: should output json asset to static folder, imported by `new URL`
### Version

  System:
    OS: macOS 14.6.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 185.28 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Browsers:
    Chrome: 129.0.6668.90
    Chrome Canary: 131.0.6763.0
    Safari: 17.6
  npmPackages:
    @rsbuild/core: ^1.0.1 => 1.0.10 
    @rsbuild/plugin-react: ^1.0.1 => 1.0.3 

### Details

If a json file is imported in the way of `asset` e.g: `new URL`, it will not be placed in the static directory, so it won't be uploaded to CDN and cause the error.

```ts
import './App.css';

const testJson = new URL('./assets/test.json', import.meta.url).href;
```

actual

```
./dist
├── 0e97e3ce1fa8f304.json
├── index.html
└── static
    ├── css
    │   └── index.e56a3ec6.css
    └── js
        ├── index.566b5367.js
        └── lib-react.a7f62b4e.js
```

expected

```
./dist
├── index.html
└── static
    ├── css
    │   └── index.e56a3ec6.css
    ├── js
    │   ├── index.566b5367.js
    │   └── lib-react.a7f62b4e.js
    └── json
        └── 0e97e3ce1fa8f304.json
```

### Reproduce link

https://github.com/sooniter/rsbuild-dist-json-issue

### Reproduce Steps

1. pnpm install
2. pnpm run build

### PR DESCRIPTION
**Pull request by SoonIter**
## Summary

feat(core): add `distPath.assets`  and `output.filename.assets`

## Related Links

close #3650

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

**Comment by SoonIter** [comment]
😂 Strictly speaking,  images svgs etc are also called `assets`.

File: packages/core/tests/__snapshots__/builder.test.ts.snap [hunk 1]
```diff
@@ -162,7 +162,7 @@ exports[`should use rspack as default bundler > apply rspack correctly 1`] = `
         "oneOf": [
           {
             "generator": {
-              "filename": "static/image/[name].[contenthash:8][ext]",
+              "filename": "static/svg/[name].[contenthash:8].svg",
```
Comment (chenjiahan) [inline packages/core/tests/__snapshots__/builder.test.ts.snap:165]: Is there any reason for this PR to change the order of the asset rules?

**Review by chenjiahan — COMMENTED** [review]
Need to update the documentation, we can create another PR to add it.

File: packages/core/tests/__snapshots__/builder.test.ts.snap (hunk 1 above)
Comment (SoonIter) [reply packages/core/tests/__snapshots__/builder.test.ts.snap]: put the easy-to-hit and high-frequency rule behind (svg, image), which would be hit as early as possible and have better performance...
small change

File: packages/core/src/plugins/asset.ts [hunk 2]
```diff
@@ -108,18 +108,19 @@ export const pluginAsset = (): RsbuildPlugin => ({
 
       const { emitAssets } = config.output;
 
-      // svg
-      createAssetRule('svg', ['svg'], emitAssets);
-      // font
-      createAssetRule('font', FONT_EXTENSIONS, emitAssets);
-      // image
-      createAssetRule('image', IMAGE_EXTENSIONS, emitAssets);
+      // put the easy-to-hit and high-frequency rule behind (svg, image), which would be hit as early as possible and have better performance
```
Comment (chenjiahan) [inline packages/core/src/plugins/asset.ts:111]: As we do not use top-level `oneof`, all top-level rules are matched and changing the rule order should have no effect on the build performance.

Comment (SoonIter) [reply packages/core/src/plugins/asset.ts:111]: maybe svg files will hit the rule eailer than before, isn't it?

**Comment by chenjiahan** [comment]
> 😂 Strictly speaking, images svgs etc are also called assets.

Haha, all of them are assets, we have categorized the assets into subdirectories, which isn't bad.

File: packages/core/src/plugins/asset.ts (hunk 2 above)
Comment (chenjiahan) [reply packages/core/src/plugins/asset.ts:111]: All rules are still to be matched...

File: website/docs/en/guide/basic/static-assets.mdx [hunk 3]
```diff
@@ -32,14 +32,22 @@ import logo from './static/logo.png';
...
 ```tsx
 import logo from '@/static/logo.png';
 
 export default = () => <img src={logo} />;
 ```
 
+You can also combine [URL](https://developer.mozilla.org/en-US/docs/Web/API/URL) with [import.meta.url](https://developer.mozilla) which are native ESM features to import static assets.
+
+```tsx
+const logo = new URL('./static/logo.png', import.meta.url).href
```
Comment (chenjiahan) [inline website/docs/en/guide/basic/static-assets.mdx:46]:
```suggestion
const logo = new URL('./static/logo.png', import.meta.url).href;
```

File: website/docs/zh/guide/basic/static-assets.mdx [hunk 4]
```diff
@@ -32,14 +32,22 @@ import logo from './static/logo.png';
...
 ```tsx
 import logo from '@/static/logo.png';
 
 export default = () => <img src={logo} />;
 ```
 
+也支持使用 ESM 原生功能中的 [URL](https://developer.mozilla.org/docs/Web/API/URL) 和 [import.meta.url](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/import.meta) 相配合，来引用静态资源。
+
+```tsx
+const logo = new URL('./static/logo.png', import.meta.url).href
```
Comment (chenjiahan) [inline website/docs/zh/guide/basic/static-assets.mdx:46]:
```suggestion
const logo = new URL('./static/logo.png', import.meta.url).href;
```

File: website/docs/zh/config/output/filename.mdx [hunk 5]
```diff
@@ -55,6 +57,7 @@ const prodDefaultFilename = {
 - `font`：表示字体文件的名称。
 - `image`：表示非 SVG 图片的名称。
 - `media`：表示视频等媒体资源的名称。
+- `assets`：表示除了上述类型之外，其他命中 `asset/resource` 的文件名称。
```
Comment (chenjiahan) [inline website/docs/zh/config/output/filename.mdx:60]: How about linking to the Asset Module documentation?

The same as `output.assetModuleFilename`

image


```suggestion
- `assets`：表示除了上述类型之外，其他命中 [Asset Modules](https://rspack.dev/guide/features/asset-module) 规则的文件名称。
```

File: website/docs/zh/config/output/dist-path.mdx [hunk 6]
```diff
@@ -51,6 +53,7 @@ const defaultDistPath = {
 - `wasm`：表示 WebAssembly 文件的输出目录。
 - `image`：表示非 SVG 图片的输出目录。
 - `media`：表示视频等媒体资源的输出目录。
+- `assets`：表示除了上述类型之外，其他也命中 [Asset Modules](https://rspack.dev/guide/features/asset-module) 的输出目录。
```
Comment (chenjiahan) [inline website/docs/zh/config/output/dist-path.mdx:56]:
```suggestion
- `assets`：表示除了上述类型之外，其他也命中 [Asset Modules](https://rspack.dev/guide/features/asset-module) 的文件的输出目录。
```

**Review by chenjiahan — APPROVED** [review]
👍🏻

**Pull request merged**